hedera-node/docs/design/services/smart-contract-service/RedirectContract-description.md

```diff
@@ -0,0 +1,263 @@
+system contract, passing back whatever the return status and value is.
+
+Expanding the template is done in
+`DispatchingEvmFrameState.get{Token,Account,Schedule}RedirectCode()`.
+
+## Raw Bytecode
+
+Derived from [`DispatchingEvmFrameState` at `e70582e28e`](https://github.com/hashgraph/hedera-services/blob/f023153207482730f79fd5da69c6fbd29eaa3fcc/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/HederaEvmWorldStateTokenAccount.java#L35) (as viewed on 2024-12-08)
+
+```
+6080604052348015600f57600080fd5b50600061RRRR905077SSSSSSSSfefefe
+fefefefefefefefefefefefefefefefefe600052366000602037600080366018
+016008845af43d806000803e8160008114605857816000f35b816000fdfea264
+6970667358221220d8378feed472ba49a0005514ef7087017f707b45fb9bf56b
+b81bb93ff19a238b64736f6c634300080b0033
+```
+
+Where:
+
+* `RRRR` will be replaced by the (2 byte) precompile contract address
+
+  * `0167`: HTS ([HIP-218](https://hips.hedera.com/hip/hip-218))
+
+  * `016a`: HAS ([HIP-906](https://hips.hedera.com/hip/hip-906))
+
+  * `016b`: HSS ([HIP-755](https://hips.hedera.com/hip/hip-755))
+
+* `SSSSSSSS` will be replaced by the (4 byte) redirect method selector
+
+  * `618dc65e`: `redirectForToken`
+
+  * `e4cbd3a7`: `redirectForAddress`
+
+  * `5c3889ca`: `redirectForScheduleTxn`
+
+* `fefefefefefefefefefefefefefefefefefefefe` will be replaced by the (20 byte) EVM address of the token
+  being redirected, i.e., the HTS token kind's account-num alias (aka: long-zero).
+
+(The templating in the source code isn't exactly organized this way, but this is the effect.)
+
+## Solidity
+
+Via `library.dedaub.com/decompile`:
+
+```solidity
+function __function_selector__() public nonPayable {
+    CALLDATACOPY(32, 0, msg.data.length);
+    v0 = 0x167.delegatecall(0x618dc65efefefefefefefefefefefefefefefefefefefefe).gas(msg.gas);
+    require(v0 != 0, 0, RETURNDATASIZE()); // checks call status, propagates error data on error
+    return MEM[0:RETURNDATASIZE()];
+}
+```
+
+## What it looks like to the calling contract
+
+Consider an example using HTS:
+
+* There's a (Hedera native) HTS token available, say at address `0.0.A`.
+* There's some address of interest, say `0xNNNNNNNNNNNNNNNNNNNN`
+* Contract thus has `0.0.A` as the address of some token, thinking it is an ERC-20 token, and has
+  the address `0xNNNNNNNNNNNNNNNNNNNN`, and it wants to know how much of that token `0.0.A` that
+  address owns.
+* Sp the contract wants to call the defined ERC-20 method
+  `balanceOf(address) external view returns (uint256)` on that token, so it:
+* Does a contract call to _the token_, at address `0.0.A`, using
+  selector [`0x70a08231`](https://www.4byte.directory/signatures/?bytes4_signature=0x70a08231),
+  passing the address of interest, `0xNNNNNNNNNNNNNNNNNNNN`, as an argument:
+  * `(bool success, uint256 amount) = 0xA.call{gas:5000}(abi.encodeWithSignature("balanceOf(address)", 0xNNNNNNNNNNNNNNNNNNNN)`
```

Comment:
should we showcase that both `call` and `staticcall` should work? The comment below about having the ERC-20 interface, would generate a `staticcall` instead.

Reply:
That's a good idea, but relatedly:  thought we still had the issue that in the smart contracts repo these methods aren't marked `view`?

Reply:
That's only applicable to two methods, `isApprovedForAll` and `getApproved` when called directly from `0x167`, not through the proxy contract. More details here https://github.com/hashgraph/hedera-smart-contracts/issues/1116.